### ISSUE TITLE
System_Boundary and lines to Systems

### ISSUE THREAD

**Issue by Yavari** [plantuml-stdlib/C4-PlantUML#103]
I want to make simple System diagram that shows the most basic dependencies within the system and also between the system and the external systems. Several internal systems have the same external dependencies so I don't want to show them all but only show that the system boundary has dependencies to external system. You can see the system diagram without the internal relationships bellow and it looks as expected. 

Now I want to add the most important internal dependencies so I am replacing System(banking_system) with System_Boundary(banking_system, "Banking System") but the lines becomes curly and the diagram becomes hard to follow.

Do you have a suggestion on how I can do this?
Is it possible to make the internal system in a different file and then import it into the previous one. So that PlantUML first renders the system boundary and then includes it a system?

You can see my example bellow

```
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Context.puml
Person(customer, "Customer", "View Websie")
Person(sales, "Sales", "Administrates admin page")
Person_Ext(helpdesk, "Customer Service", "Form Requests")
System(banking_system, "Banking System", "Customer website, Admin website, External REST API")

System_Ext(mail_system, "E-mail system", "Exchange server")
System_Ext(ip_system, "Integration Plattofrm", "Customer data")
System_Ext(contract_system, "Contact", "Contracts")
System_Ext(s2_system, "System 2", "Information")
System_Ext(s3_system, "System 3", "Information")

Rel(customer, banking_system, "Uses")
Rel(sales, banking_system, "Uses")

Rel_Neighbor(banking_system, mail_system, "Sends e-mails", "SMTP")
Rel_Back(helpdesk, mail_system, "Sends e-mails to")

Rel(banking_system, ip_system, "Uses", "WCF")
Rel(banking_system, contract_system, "Uses", "SQL")
Rel(banking_system, s2_system, "Uses", "SQL")
Rel(banking_system, s3_system, "Uses", "SQL")
@enduml
```

![bild](https://user-images.githubusercontent.com/1221772/102684221-f4c09280-41d6-11eb-8dd4-baea703d0f79.png)

```
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Context.puml
Person(customer, "Customer", "View Websie")
Person(sales, "Sales", "Administrates admin page")
Person_Ext(helpdesk, "Customer Service", "Form Requests")

'This is the change
System_Boundary(banking_system, "Banking System"){
    System(web, "Banking System", "Customer websiteI")
    'System(admin, "Banking System", "Admin website") not even needed in example
    'System(api, "Banking System", "External REST API") not even needed in example
}

System_Ext(mail_system, "E-mail system", "Exchange server")
System_Ext(ip_system, "Integration Plattofrm", "Customer data")
System_Ext(contract_system, "Contact", "Contracts")
System_Ext(s2_system, "System 2", "Information")
System_Ext(s3_system, "System 3", "Information")

Rel(customer, banking_system, "Uses")
Rel(sales, banking_system, "Uses")

Rel_Neighbor(banking_system, mail_system, "Sends e-mails", "SMTP")
Rel_Back(helpdesk, mail_system, "Sends e-mails to")

Rel(banking_system, ip_system, "Uses", "WCF")
Rel(banking_system, contract_system, "Uses", "SQL")
Rel(banking_system, s2_system, "Uses", "SQL")
Rel(banking_system, s3_system, "Uses", "SQL")
@enduml
```

![bild](https://user-images.githubusercontent.com/1221772/102684320-b677a300-41d7-11eb-8d29-29391d6c4782.png)

**Comment by adrianvlupu** [comment]
Hi, 

Regarding

> Is it possible to make the internal system in a different file and then import it into the previous one.

You can use `!includesub`

`main.puml`
```plantuml
...

System_Boundary(banking_system, "banking System"){
    !includesub ./banking_containers.puml!containers
}

...
```

`banking_containers.puml`

```puml
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml

!startsub containers
Container(c1, "container", "tech")
!endsub

@enduml
```

![test](https://www.plantuml.com/plantuml/png/ZLHHQzim47utx3_Sl8Sbi2HRqwTHIjVGmA0TMRskZqQMBxQ8BNdIkMbMzjztift6ISYO25Zwz7tVdUwEN3aIbkgo2DugBOiwHSY9AlTvCh5YFSuKvNLIExJIQ49DOsdAILK8xJKZHscXaidiVBJmqCFTxQGKZj0o5CzO8PH6EwwyV-1MIbV2YX8iIX8wA_3YWj3HvIMy1C0hoaLgrafdi1I5mnPZDpIqgN3EVf7wZV2nkV37cIcCXNVdpMhGQsDJjAVmY3O5GYfSZcaB2xdAhAbrsadcpGh-1C42hJDw86j7faJx0S9Pjmrv_rFX6XunSGh3uPRhH86EBoF_zQmlQQcqScG5lnE4Fq4bifqaldccGOv5bQ9RzK90XFP9IVGcSsDBkCTVDPVAiJ9u3vGh1_pBpHe493v_OJTWQgfgSa6quHwKyNM73nAXkVPPx1gSBJi0Mc8uV0bUcpMGdvWWjqUV0A7Cm-4RNvJEsMTT2bNi76z6_WpjsOjkdcNEFKRWsNbYds5VgwgTyYlFLSO5KaO3Jn6HMTh6uRKMgI2nh_VfMI5fvzBaBodijlxIxMlSsOxTmd0MDl6NNE0c_e5WUYIO7Wk2UonwGt9KwWU7FX7FwiRZ5CLpuc-eiZmnzgXd--MEKASEi2bwCsdHtOz56oRsCzsRgNy8WSps1OVX-XrgykFluspUlUIGVToFhIRwVdj2q-_7Vx2dfzXN_2B-V_aB "test")

> System_Boundary(banking_system, "Banking System") but the lines becomes curly and the diagram becomes hard to follow.

I tried your examples and you are correct. When adding any type of nested rectangle the arrows break.

This looks fine:
```plantuml
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml

skinparam rectangle<<test>> {
    Shadowing false
    StereotypeFontSize 0
    FontColor #444444
    BorderColor #444444
    BorderStyle dashed
    BackgroundColor #FFFFFF
}

Person(customer, "Customer", "View Websie")
Person(sales, "Sales", "Administrates admin page")
Person_Ext(helpdesk, "Customer Service", "Form Requests")

' this is how a boundary outputs
rectangle "==Test" <<test>> as banking_system {

}

System_Ext(mail_system, "E-mail system", "Exchange server")
System_Ext(ip_system, "Integration Plattofrm", "Customer data")
System_Ext(contract_system, "Contact", "Contracts")
System_Ext(s2_system, "System 2", "Information")
System_Ext(s3_system, "System 3", "Information")

Rel(customer, banking_system, "Uses")
Rel(sales, banking_system, "Uses")

Rel_Neighbor(banking_system, mail_system, "Sends e-mails", "SMTP")
Rel_Back(helpdesk, mail_system, "Sends e-mails to")

Rel(banking_system, ip_system, "Uses", "WCF")
Rel(banking_system, contract_system, "Uses", "SQL")
Rel(banking_system, s2_system, "Uses", "SQL")
Rel(banking_system, s3_system, "Uses", "SQL")
@enduml
```

![test](https://www.plantuml.com/plantuml/png/ZPDjJzim4CVVvrCSsGj0Mjkfy6e2n5QrqYIO687nidAI8x58xCntePHfttrdDtrMfrcLOf__lplVGw-9bUEshg8ZRRAgpH5AveO-3mPEpVg5vh9DMqANMSDek9_PUj1KodYcHvnNEXsCpdjttlHuUpEe5J4wCKr7GYXjqFKRxp-Y5sqQvLGD3ZDMfgZmueAH-EeAVaSWAobLRcVQ5F2iAiA5JRoXvNc34_6Nw7U4J-727qUsiWu-d8SLh5-joz4TiYSyhn1oHINc2xFANWfdMvDto2Iiw4yKtQ4ZQqwobjZMw3v2FEgsiUn_Qfp14wQaCJvTQabLI7AP-AzNVSbhRJInKv8dA7-2HXLhP3f-uvCIgoP7Uja80GcwLvsXTpAnhePx_DLAgKZ8w1Yur0JoA-qC5AJ-_ShDmRRSj4pHghuGNru-21R3gjAA85L6Mb5CQIwjghlQ7yCMjMfUx0dFPajB9r_E2xKfz0TUCGoDFbe9oPSn2K52dhNILHTKqXht_1aMPv_d-2qh9Jw2JDkh13xTP7Mp9h_991PIKcqDoDmnssSNFAogbojMsxn_b5CPhxs433EEkws_f6s6XclrmWp3ECH_bfQ4-3l0sHvmjWz4zrXjZDLsDqJzIEWVubNTG1sIUCtqE-gYJAqxsLLjbpj1anDWA7gOpUJsuMuHPkh_1HjJ-0yGs2ups0stsQ7mFla-ZIPT9hlg_NuicUJ7pG5cin__eJuxfBwMZ6HG_m80 "test")

and adding the nested rectangle breaks the layout:

```plantuml
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml

...
rectangle "==Test" <<test>> as banking_system {
    rectangle "==Container" as testcontainer
}

...
@enduml
```

![test](https://www.plantuml.com/plantuml/png/ZLFlJzim4FsUl-B8FW3Isaw5JnCWjggL9i7428oFbPCSYKLYPxuBfKpxttTsqz_gDAjIxFDxTxxthjV4od5RLz6HDbdLvWWbSqEV1mEdPlr2SzccBQ7Bh64qtCziFMWgPJodHvnNEXsCpdjtFlHuUpEe5J4w2Kr7mb3QeEitFdz4BzeqogaQ76QiJ57XnGKZyTKL_8v0LbAgtCwqAU1PLOIBc6H3o_C69v8lqUy8dyA5FuviPHryE0yhHBzQbwCx54zuNY7aYahC5s6LlHJEjYRlA9EmeZzHT8UEh3d9Mc9Re_i8yQZRnhB_gN46JvYInlXqYILL8SbbuhyUzIMljT74JacVeFm96bMiATFn6v-KM3Kvqij62KZGlUeCVPA9TJNSuwzMf29XHiV0fIQGNsbde23rxrTk3hRbfcMALlf2V7dv8BGOLaehWbGPiQAOqboigZljjpWhwsAFzyni6GdWOrXE2xKfz0SUCGms7ws0v4LCGedGPQrqrPMKfiOzVuR5sNSvVijAgOyWi_OgXKytkRfPCx_971OYgBO6PEgOxRCB6LRQvOhLDjy_ogcCrrb2Xnd7tTPVqZQ7Xclq8Wp3EDH_5aD2_Ht2sHxXR9yGtMErCLJRNWZwaT0_nAEwSJe4yPZfTzH5cLftielQbZj1anDW43rCPdBxSBSeC_N_WOqP_0SHs2uxs2stwL1udtoVHfEkartqlXzBJlBZvW1dqu__G9yTGbzBHpAeVm40 "test")

It seems to be a PlantUML thing, found something similar to your issue here https://forum.plantuml.net/11778/relationship-layout-with-nested-components but had no luck actually generating a diagram using the sub-diagram.

**Comment by Yavari** [comment]
Thank you for taking the time to look at this.
The issue you found at plantuml forum looked promising and it works when I run it on the server ([view here](http://www.plantuml.com/plantuml/uml/ROn12i9034NtEKKkq4AxhrJFu4vSX9bOIZEfP38WbDxTIWL5t8JFo-FrsScy984yZNedeuGd2jDWSz4OPfaDtKZpzcBrBoloZOhucHz-b77GnEeOjiC6O1oST136rshm0bstJwXhM1P0_2ZAsJasVrWBwmhNdreBi4ShgmvlqhF6akG9)). It doesn't work from Visual Studio Code but I can live with that for now.

Exactly the same code but with the C4_Container added throws an exception ([view here](http://www.plantuml.com/plantuml/uml/ROt1IaCn44NtynNn0rv2YokXKkrMmOMkH6IQZA-XaybZCa6XzD_jeu8YtLtEt3bttGpLUc60gooHUoAtCvlQJGYAdtxCjklRtaXZ5ICn7si94wFCJqEpn7aRDjV3qunU7Xz2mMQa9_I-gMAOXTHFyq3RPvbGiRXxZFjHQvSKArTrfYZjT2AnFwr47zZPdkdBxZYFKaZCHH8Z1QLeA2EJczF2lS9gTOxWFHmEuDnlHQWPfSK5jeJZ4Tx-sPO0P-cWm-qFMPEaNlWR)):
```
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml
```

**Comment by Yavari** [comment]
I have found that it is line 17 in C4.puml that is casing the issue:
skinparam wrapWidth 200

Is it possible to define a type for the C4 elements and set the wrapWidth attribute on that type instead? 

```
@startuml
skinparam wrapWidth 200

rectangle rect1 [
== rect1
..
{{
  rectangle nested1
  rectangle nested2
}}
]
rectangle rect2

rect1 -r-> rect2
@enduml
```

**Comment by adrianvlupu** [comment]
Interesting problem, you cound fix it by adding `skinparam wrapWidth auto` after the include but all the container/system entities will need to use `\n`

```plantuml
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml

skinparam wrapWidth auto

skinparam Backgroundcolor transparent
skinparam defaultTextAlignment center
rectangle rect1 [
== rect1
..
{{
  rectangle nested1
  rectangle nested2

Container(c1, "Container", "tech", "Some long description of the container.\n You have to use new line characters\n in order to keep the layout")
}}
]
rectangle rect2

rect1 -r-> rect2
@enduml
```

![image](http://www.plantuml.com/plantuml/png/RL1VIoCn47_FfvZhqntSxjBZdeHAjQyAWeg8YaoJwMveTh9C9bOf_UvchTGYlYGplp_pP-P9KJJtmVpoR4Ds19tga4wQHd1JjrwxlCo9n4PMOgrjx9iX88-UAgaBVjai_bTN8tHxUT7qc9IaGC-Bua3F9FKmrZTfxNb0mHusWiETTze1Pet7p3dQTIinixCnH04Lv5GeOZrIELfX3df3hteMVCjzeS6MXyG8MKLk0y4OJU71p6RxqDIrsMuDm9U4AIcvwG_OFsCEy_-sqxymEQIJaYZPRlolOqyG8hVWA5dnW_h845UW7O4z7E2HuJvcwF25G2FaDBRPG2Was0u5hPAaelB5B8vaLAs9Xeyw0TzYriaVizkPfsyBbZdtUrPIdNuYSs9NBluE)

**Comment by Yavari** [comment]
Thanks, that works for me.
